// @walletconnect/utils/dist/cjs/crypto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decrypt = exports.encrypt = exports.sha256 = exports.deriveSharedKey = exports.generateRandomBytes32 = exports.generateKeyPair = void 0;
const tslib_1 = require("tslib");
const ecies25519 = tslib_1.__importStar(require("@walletconnect/ecies-25519"));
const encoding = tslib_1.__importStar(require("@walletconnect/encoding"));
function generateKeyPair() {
    const keyPair = ecies25519.generateKeyPair();
    return {
        privateKey: encoding.arrayToHex(keyPair.privateKey),
        publicKey: encoding.arrayToHex(keyPair.publicKey),
    };
}
exports.generateKeyPair = generateKeyPair;
function generateRandomBytes32() {
    return encoding.arrayToHex(ecies25519.randomBytes(32));
}
exports.generateRandomBytes32 = generateRandomBytes32;
function deriveSharedKey(privateKeyA, publicKeyB) {
    const sharedKey = ecies25519.derive(encoding.hexToArray(privateKeyA), encoding.hexToArray(publicKeyB));
    return encoding.arrayToHex(sharedKey);
}
exports.deriveSharedKey = deriveSharedKey;
function sha256(msg) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const hash = yield ecies25519.sha256(encoding.hexToArray(msg));
        return encoding.arrayToHex(hash);
    });
}
exports.sha256 = sha256;
function encrypt(params) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const msg = encoding.utf8ToArray(params.message);
        const sharedKey = encoding.hexToArray(params.sharedKey);
        const publicKey = encoding.hexToArray(params.publicKey);
        const iv = typeof params.iv !== "undefined" ? encoding.hexToArray(params.iv) : undefined;
        const encrypted = yield ecies25519.encryptWithSharedKey(msg, sharedKey, publicKey, iv);
        return encoding.arrayToHex(encrypted);
    });
}
exports.encrypt = encrypt;
function decrypt(params) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const encrypted = encoding.hexToArray(params.encrypted);
        const sharedKey = encoding.hexToArray(params.sharedKey);
        const msg = yield ecies25519.decryptWithSharedKey(encrypted, sharedKey);
        return encoding.arrayToUtf8(msg);
    });
}
exports.decrypt = decrypt;
