// @walletconnect/utils/dist/cjs/uri.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseUri = exports.formatUri = void 0;
const tslib_1 = require("tslib");
const qs = tslib_1.__importStar(require("query-string"));
const safe_json_1 = require("@walletconnect/safe-json");
function formatUri(params) {
    return (`${params.protocol}:${params.topic}@${params.version}?` +
        qs.stringify({
            publicKey: params.publicKey,
            controller: params.controller,
            relay: safe_json_1.safeJsonStringify(params.relay),
        }));
}
exports.formatUri = formatUri;
function parseUri(str) {
    const pathStart = str.indexOf(":");
    const pathEnd = str.indexOf("?") !== -1 ? str.indexOf("?") : undefined;
    const protocol = str.substring(0, pathStart);
    const path = str.substring(pathStart + 1, pathEnd);
    const requiredValues = path.split("@");
    const queryString = typeof pathEnd !== "undefined" ? str.substr(pathEnd) : "";
    const queryParams = qs.parse(queryString);
    const result = {
        protocol,
        topic: requiredValues[0],
        version: parseInt(requiredValues[1], 10),
        publicKey: queryParams.publicKey,
        controller: queryParams.controller === "true",
        relay: safe_json_1.safeJsonParse(queryParams.relay),
    };
    return result;
}
exports.parseUri = parseUri;
